### DISCUSSION TITLE
Current best option for iPad USB C Raspberry Pi?

### DISCUSSION THREAD

**Discussion by c0ppelius** [coder/code-server#2940] [Q&A] [answered]
I have an iPad Pro and a Raspberry Pi running 64Bit Rasbian. I’ve connected the Pi to the iPad via an Ethernet USB C connection. I would like to securely expose code-server via the Ethernet connection alone while using WiFI on on the Pi. One option seems to be using an ssh client on the iPad to port forward the connection. The apparent drawback is a broken connection when the ssh app is backgrounded. Otherwise, it should work. 

Another option that suffers from the same drawback is Servediter. 

Another option that should almost get me there is setting up mDNS via avahi to get a domain name and using a self signed certificate. As far as I understand it, this would expose code-server on the local network connected to the WiFi. 

Another possible option is to use something like raspap to have a hotshot on the Pi and forget about the USB connection though I haven’t though this through completely. 

Am I making this too complicated? What experiences have people had and what resources have I missed?

My main use case would be the Pi connected to an untrusted LAN via WiFi (say while traveling).

**Comment by jsjoeio** [comment]
Would `--link` work here? see [video](https://www.youtube.com/watch?v=MdIEiZpg5vQ). I'm guessing you've read the iPad docs, but if not, [link here](https://github.com/cdr/code-server/blob/main/docs/ipad.md).

@bpmct might be able to help

**Reply by c0ppelius** [reply]
Thanks for pointing this out. The `--link` flag would be another solution when there is an external connection for both the Pi and iPad - no USB C needed. I should have mentioned another use case is on a plane (someday). So ideally I would like something that also works when the only connection is the USB C ethernet.  

I noticed that docs don't mention the possibility of ssh port forwarding on the iPad. This seems feasible using one of the standard ssh apps on the iPad, e.g. [Blink](https://github.com/blinksh/blink/issues/142). 

I haven't seriously tried ssh port forwarding. So far I've just messed around with caddy or nginx over the USB C where I ran into the standard white blank page after the login. I thought it would be a good idea to try to collect knowledge about the possible avenues before taking another run at it. You should assume I know nothing.

**Reply by jsjoeio** [reply]
Ooo I didn't know Blink was a thing. $20 too 😂 But hey, super cool to see someone built that. 

> You should assume I know nothing.

Same for me lol But in all seriousness, I haven't tried ssh port forwarding on the iPad yet. I've only taken the easy routes: `--link`, Railway, etc. (see the deploy code-server [here](https://github.com/cdr/deploy-code-server), Railway works decently on iPad but wouldn't solve the plane scenario).

Definitely out of my wheelhouse, but maybe Ben can chime in when he's back from vacation.

**Reply by c0ppelius** [reply]
[iSH](https://ish.app/) is free (I think) and might be the best option between it, Blink, and Terminus. In fact, there is [internet confirmation](https://www.reddit.com/r/ipad/comments/khd9fm/codeserver_via_ssh_port_forwarding_works_with_the/) of getting ssh forwarding working. My main concern is losing connection either due to iOS backgrounding the ssh app or the iPad heading into sleep. But perhaps that is not a big concern. :shrug:

**Reply by jsjoeio** [reply]
TIL about iSH - thanks for the link! Now I want to try this out. 

> My main concern is losing connection either due to iOS backgrounding the ssh app or the iPad heading into sleep

Well, I can't say this applies to you because I don't know what you're planning to build on your iPad, but in my experience, I haven't really had to worry about this! But most of the time, my workflow is:
- open code-server on iPad
- write some code
- commit
- close iPad

Don't know if that helps 🤷‍♂️

**Reply by c0ppelius** [reply]
You’ve definitely helped. I’ve run code-server with the remote machine in the cloud with my local machine being the iPad. It is totally robust in my experience. The ssh app however has been much more problematic. Usually, if I want a stable connection, I mosh in instead of ssh.

**Comment by c0ppelius** [comment, marked as answer]
~~As I feared, the ssh connection via an iPad ssh app drops pretty quickly (within 10-15 seconds) when the app is backgrounded. (Tested with both iSH and Blink using the geo track option.) Split screen seems to be fine in terms of stability (so far).~~

I was not using the `geo lock` command correctly on Blink. Supplying a reasonable radius for the geo lock gives a robust connection that is stable even after the iPad sleeps. 

Quick summary of what I've done:
- Installed 64bit Rasbian to have access to the code-server .deb package from [here](http://downloads.raspberrypi.org/raspios_arm64/images/)
- Set up the USB-C connection for ethernet as described [here](https://medium.com/sausheong/setting-up-a-raspberry-pi-4-as-an-development-machine-for-your-ipad-pro-3813f872fccc)
- Run the install script for code-server and followed the instructions on the [guide](https://github.com/cdr/code-server/blob/main/docs/guide.md#ssh-forwarding) to set up ssh port forwarding. 
- Installed Blink (or comparable iOS app which allows geographical tracking to run in the background). Run `geo lock Nm` where `Nm` is the radius in meters you are locked to and then run `ssh -L 8080:localhost:8080 xx.xx.xx.xx` where `xx.xx.xx.xx` is the IP address of your Raspberry Pi 
- Navigated to localhost:8080 and enjoyed code-server running via a stable USB-C ethernet connection between the "remote" raspberry pi and local iPad.

**Reply by jsjoeio** [reply]
Woohoo! Glad to hear it sounds like you found a solution. Feel free to mark it as the answer!

**Reply by oxy** [reply]
I haven't used Blink or similar on an iPad, but I'm curious - would it be possible to just expose `code-server` directly from the RPi (with `--bind-addr 0.0.0.0:8080`) and then just connect to it via Safari with `xx.xx.xx.xx:8080` directly?

That should solve the port-forwarding problem - obviously then anyone who is on the network can connect too, but since you mentioned a USB C Ethernet setup, I think that should be a non-issue.

**Reply by c0ppelius** [reply]
Wouldn’t this expose code-server over all interfaces, here both the Ethernet and WiFi connections? Is there a flag to specify the interface when calling code-server?

Regarding ssh port forwarding, the ssh connection (via WiFi on the LAN) was completely stable overnight. So I think enabling location services on the ssh app does work to keep from iPadOS dropping the connection. 

Before marking anything as an answer, I want to see who else shows up with ideas. I’ll give it a week or so. Hopefully, I’ll learn something new. Here is a main downside to the procedure above: as written it requires a $20 app to implement.
